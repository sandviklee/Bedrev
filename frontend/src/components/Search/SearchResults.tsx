import { gql, useQuery } from "@apollo/client";
import { Company } from "../../types/types";
import CompanyCard from "../Company/CompanyCard";
import { Link } from "react-router-dom";
import Pagination from "@mui/material/Pagination";
import { useEffect, useState } from "react";
import {
  minValueState,
  placeState,
  maxValueState,
  searchState,
  sortState,
} from "../../atoms/atoms";
import { useRecoilState } from "recoil";
import { useDebounce } from "use-debounce";

const SearchResults = () => {
  const [currentPage, setCurrentPage] = useState(1);
  const place = useRecoilState<string>(placeState);
  const searchValue = useRecoilState<string>(searchState);
  const minValue = useRecoilState<string>(minValueState);
  const maxValue = useRecoilState<string>(maxValueState);
  const sort = useRecoilState<string>(sortState);
  const [minDebounced] = useDebounce(minValue[0], 500);
  const [maxDebounced] = useDebounce(maxValue[0], 500);
  const [searchDebounced] = useDebounce(searchValue[0], 500);

  /**
   * Updates the session storage.
   */
  useEffect(() => {
    sessionStorage.setItem("minValue", minDebounced);
    sessionStorage.setItem("maxValue", maxDebounced);
    sessionStorage.setItem("sort", sort[0]);
    sessionStorage.setItem("place", place[0]);
    sessionStorage.setItem("search", searchDebounced);
  }, [minDebounced, maxDebounced, sort[0], searchDebounced, place[0]]);

  const BEDRIFT_QUERY_SEARCH = gql`
        {
            bedriftSok(skip: ${
              currentPage > 0 ? (currentPage - 1) * 5 : 0
            }, navn: "${searchDebounced
              .toLowerCase()
              .trim()}", lavest:${parseInt(minDebounced)}, hoyest:${parseInt(
              maxDebounced,
            )}, postadresse_poststed: "${place[0].toUpperCase()}", sort: "${
              sort[0]
            }") {
                navn
                postadresse_kommune
                antallansatte
                vedtektsfestetformaal
                organisasjonsnummer
            }
            bedriftSokCount(navn: "${searchDebounced
              .toLowerCase()
              .trim()}", lavest:${parseInt(minDebounced)}, hoyest:${parseInt(
              maxDebounced,
            )}, postadresse_poststed: "${place[0].toUpperCase()}")
        }
    `;

  const { data, loading, error } = useQuery(BEDRIFT_QUERY_SEARCH, {
    fetchPolicy: "cache-first", // Used for first execution
    nextFetchPolicy: "cache-first", // Used for subsequent executions
  });

  /**
   * Handler for the pagination component.
   * @param event change of page
   * @param value is the current page value
   */
  const handleChange = (event: React.ChangeEvent<unknown>, value: number) => {
    event;
    setCurrentPage(value);
  };

  /**
   * Resets the page to the first page everytime
   * the search result is filtered.
   */
  const resetPage = () => {
    setCurrentPage(1);
  };
  useEffect(() => {
    resetPage();
  }, [searchValue[0], minDebounced, maxDebounced, place[0]]);

  if (loading)
    return <p className="text-left text-lg">Laster inn flere bedrifter...</p>;
  if (error)
    return (
      <>
        <div className="leading-[30px] py-[20px]">
          <p className="font-medium tracking-tight">
            Beklager, vi klarer ikke å laste inn noen bedrifter akkurat nå, prøv
            igjen senere :(
          </p>
          <p className="font-bold tracking-wider">
            Feilmelding: {error.message}
          </p>
        </div>
      </>
    );

  return (
    <>
      <p className="font-inter tracking-wide sm:text-[14px] text-[12px] ">
        Fant {data["bedriftSokCount"]} bedrifter med valgte filter
      </p>
      <div className="flex flex-col sm:gap-1 gap-3 mt-4">
        {data["bedriftSok"].map((company: Company) => (
          <Link key={company.navn} to={String(company.organisasjonsnummer)}>
            <CompanyCard {...company} />
          </Link>
        ))}
      </div>
      <div className="flex justify-center pt-8">
        <Pagination
          page={currentPage}
          onChange={handleChange}
          count={Math.ceil(data["bedriftSokCount"] / 5)}
        />
      </div>
    </>
  );
};

export default SearchResults;
