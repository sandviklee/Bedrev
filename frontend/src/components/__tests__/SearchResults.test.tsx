import { render, screen, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { gql } from "@apollo/client";
import SearchResults from "../Search/SearchResults";
import { BrowserRouter as Router } from "react-router-dom";
import { RecoilRoot, MutableSnapshot } from "recoil";
import {
  placeState,
  searchState,
  minValueState,
  maxValueState,
  sortState,
} from "../../atoms/atoms";

const getBedriftQuerySearch = (
  currentPage: number,
  searchValue: string,
  minDebounced: string,
  maxDebounced: string,
  place: string,
  sort: string,
) => gql`
    {
        bedriftSok(skip: ${
          currentPage > 0 ? (currentPage - 1) * 5 : 0
        }, navn: "${searchValue.toLowerCase().trim()}", lavest:${parseInt(
          minDebounced,
        )}, hoyest:${parseInt(
          maxDebounced,
        )}, postadresse_poststed: "${place.toUpperCase()}", sort: "${sort}") {
            navn
            postadresse_kommune
            antallansatte
            vedtektsfestetformaal
            organisasjonsnummer
        }
        bedriftSokCount(navn: "${searchValue
          .toLowerCase()
          .trim()}", lavest:${parseInt(minDebounced)}, hoyest:${parseInt(
          maxDebounced,
        )}, postadresse_poststed: "${place.toUpperCase()}")
    }
`;

const mocks = [
  {
    request: {
      query: getBedriftQuerySearch(1, "test", "0", "100", "PLACE", "sort"),
    },
    result: {
      data: {
        bedriftSok: [
          {
            navn: "Test Company",
            postadresse_kommune: "Oslo",
            antallansatte: 508,
            vedtektsfestetformaal: "Lage ting",
            organisasjonsnummer: 123456789,
          },
        ],
        bedriftSokCount: 1,
      },
    },
  },
];

const initializeState = (mutableSnapshot: MutableSnapshot) => {
  mutableSnapshot.set(placeState, "PLACE");
  mutableSnapshot.set(searchState, "test");
  mutableSnapshot.set(minValueState, "0");
  mutableSnapshot.set(maxValueState, "100");
  mutableSnapshot.set(sortState, "sort");
};

describe("SearchResults component", () => {
  test("should render companies", async () => {
    render(
      <Router>
        <RecoilRoot initializeState={initializeState}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <SearchResults />
          </MockedProvider>
        </RecoilRoot>
      </Router>,
    );

    await waitFor(() => screen.getByText("TEST COMPANY"));

    expect(screen.getByText("TEST COMPANY")).toBeInTheDocument();
    expect(screen.getByText("Oslo")).toBeInTheDocument();
    expect(screen.getByText("508")).toBeInTheDocument();
    expect(screen.getByText("Lage ting")).toBeInTheDocument();
  });

  test("should change page when pagination is clicked", async () => {
    render(
      <Router>
        <RecoilRoot initializeState={initializeState}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <SearchResults />
          </MockedProvider>
        </RecoilRoot>
      </Router>,
    );

    await waitFor(() => screen.getByText("1"));

    screen.getByText("1").click();

    expect(screen.getByText("1")).toHaveClass("Mui-selected");
  });

  test("matches snapshot", async () => {
    const { asFragment } = render(
      <Router>
        <RecoilRoot initializeState={initializeState}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <SearchResults />
          </MockedProvider>
        </RecoilRoot>
      </Router>,
    );

    await expect(asFragment()).toMatchSnapshot();
  });
});
