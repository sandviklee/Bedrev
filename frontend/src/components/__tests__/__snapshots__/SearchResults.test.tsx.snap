// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchResults component > matches snapshot 1`] = `
<DocumentFragment>
  <p
    class="text-left text-lg"
  >
    Laster inn flere bedrifter...
  </p>
</DocumentFragment>
`;
